model/task/task.go

```diff
@@ -2651,44 +2651,147 @@ func (t *Task) MarkUnscheduled() error {
-// returns all the tasks after the update.
-func MarkAllForUnattainableDependency(ctx context.Context, tasks []Task, dependencyID string, unattainable bool) ([]Task, error) {
+// MarkAllForUnattainableDependency updates many tasks (taskIDs) to mark a
+// subset of all their dependencies (dependencyIDs) as attainable or not. If
+// marking the dependencies unattainable, it creates an event log for each newly
+// blocked task. This returns all the tasks after the update.
+func MarkAllForUnattainableDependencies(ctx context.Context, tasks []Task, dependencyIDs []string, unattainable bool) ([]Task, error) {
 	if len(tasks) == 0 {
 		return tasks, nil
 	}
 
+	dependencyIDSet := map[string]struct{}{}
+	for _, depID := range dependencyIDs {
+		dependencyIDSet[depID] = struct{}{}
+	}
+
+	toUpdate := getTaskDependenciesToUpdate(tasks, dependencyIDSet, unattainable)
+
+	if err := updateTaskDependenciesChunked(ctx, toUpdate, unattainable); err != nil {
+		return nil, errors.Wrap(err, "updating task dependencies in chunks")
+	}
+
+	event.LogManyTasksBlocked(ctx, toUpdate.newlyBlockedTaskData)
+
+	updatedTasks, err := FindAll(db.Query(ByIds(toUpdate.taskIDs)))
+	if err != nil {
+		return nil, errors.Wrap(err, "finding updated tasks")
+	}
+	if len(updatedTasks) != len(tasks) {
+		return nil, errors.Errorf("updated dependencies for tasks but subsequent query for updated tasks returned a different number of tasks (%d) than expected (%d)", len(updatedTasks), len(tasks))
+	}
+
+	return updatedTasks, nil
+}
+
+type taskDependencyUpdate struct {
+	// taskIDs is the list of task IDs whose dependencies should be updated.
+	taskIDs []string
+	// taskDependenciesToUpdate is a mapping of task IDs that need updating to
+	// the dependency IDs that should be updated for that task.
+	taskDependenciesToUpdate map[string][]string
+	// newlyBlockedTaskData contains info on tasks that were not blocked but are
+	// about to be blocked.
+	newlyBlockedTaskData []event.TaskBlockedData
+}
+
+func getTaskDependenciesToUpdate(tasks []Task, dependencyIDSet map[string]struct{}, unattainable bool) taskDependencyUpdate {
 	taskIDs := make([]string, 0, len(tasks))
+	taskDependenciesToUpdate := make(map[string][]string)
 	newlyBlockedTaskData := make([]event.TaskBlockedData, 0, len(tasks))
 	for _, t := range tasks {
-		if !t.Blocked() && unattainable && !t.OverrideDependencies {
+		taskIDs = append(taskIDs, t.Id)
+
+		taskCanBeNewlyBlocked := unattainable && !t.Blocked() && !t.OverrideDependencies
+		var taskBlockedOn string
+		for _, dependsOn := range t.DependsOn {
+			depID := dependsOn.TaskId
+			if _, ok := dependencyIDSet[depID]; !ok {
+				continue
+			}
+
+			taskDependenciesToUpdate[t.Id] = append(taskDependenciesToUpdate[t.Id], depID)
+
+			if taskCanBeNewlyBlocked && taskBlockedOn == "" {
+				// If a task is going to be newly blocked by multiple
+				// dependencies at once, just block it arbitrarily on
```

Comment:
Is sthis information just used for logging? Can we mention that we're just logging it as blocked arbitrarily? Seems like we should still mark the task blocked for all the dependencies so we don't end up in a weird state (I might be reading too fast)

Reply:
This should be the same as it was before, just being done in bulk instead of one at a time. The way the old algorithm worked is it would find the first dependency that switched the task from unblocked to blocked, and then log one task blocked event for it. If later tasks in the dependency graph also would keep the task blocked, it wouldn't log those.

I imagine this is probably just because there were too many task blocked logs to reasonably log every occurrence and it was already blocked anyways, so later blocks aren't actually changing the task's blocked state.